### ISSUE TITLE
Understand the mass difference between our and WTG estimates for MACSJ

### ISSUE THREAD

**Issue by nicolaschotard** [nicolaschotard/Clusters#21]
Here are the things that we need to test to understand where is this difference coming from. To do so, two cases have been identified for the CFHT vs. WTG catalogs comparison. 

- Matched catalogs: keep the objects present in both catalogs for these comparisons
  - how many galaxies do we have in common?
  - compare ZBEST from a standard BPZ run
  - compare shape estimates through the shear profile plot
      - average shear as a function of distance to the cluster center)
      - use the same band for this comparison
- Unmatched catalogs: keep all galaxies from both samples and make statistical comparisons
  - how many galaxies do we have in our sample after the radius cut applied before the mass estimate?
  - check and list all our cuts
  - compare magnitude distributions
  - compare the Delta Sigma profiles
  - prior on the mass is different in the two analyses, and here is the effect on changing the prior in our analysis (results from Douglas)
    - linear prior: mean M200 = 5.3 x 10^14
    - log prior: mean M200 = 2.8 x 10^14 
  - cluster contamination in our sample? try to change the cut on redshift
  - no shear calibration applied in our analysis. See what we can do about that

Another obvious thing to test is to run `pzmassfitter` on the WTG data (redhift PDZ and shear). We should get a result ery close to the published value.

**Comment by deapplegate** [comment]
I'll need to re-add options to alter the prior, otherwise running the PZMassFitter on the WTG data will be different than published. I'll have the pzmassfitter module read options out of the configuration file instead of parsing the command line.

The published values used a prior that was ln M(<1.5MPC) ~ Uniform(ln(10^13), ln(10^16) I think. The current prior is M_200 ~ Uniform(10^13, 10^16).

**Comment by combet** [comment]
Just pushed `scripts/clusters_import_wtgcat.py`. It gathers information from the various WTG files (shear, zbest, pdz) and produces a pipeline-compatible `cat_wtg.hdf5` file that can be processed directly by `clusters_mass`.

Path to the WTG catalog files and output hdf5 file should be changed "manually" in the script.

Need to confirm with @deapplegate that the right fields from the WTG catalogs are used.

**Comment by deapplegate** [comment]
Could you post the fields used?

**Comment by combet** [comment]
* `gs1` and `gs2` from MACS2243-09.W-J-V.gab1728-rot1.cut_lensing.cat
* `BPZ_Z_B` from MACS2243-09.W-J-V.bpz.tab
* `pdz` from MACS2243-09.W-J-V.pdz.cat
* `zrange = np.arange(header['MINPDZ'],header['MAXPDZ'],header['PDZSTEP'])`

Common objects between files identified using `SeqNr`

**Comment by deapplegate** [comment]
Code for logM prior, substitue into maxlike_masses.py lines approx. 208 - 212:

      ```
  parts.logmdelta = pymc.Uniform('logmdelta', np.log(options.masslow),
                                       np.log(options.masshigh))

 

 
        @pymc.deterministic
        def mdelta(logmdelta = parts.logmdelta):
            return np.exp(logmdelta)
        parts.mdelta = mdelta

```

**Comment by deapplegate** [comment]
Re-emphasize that the HSM shape shear calibration needs to be included if we want DMStack masses to match WtG.

**Comment by deapplegate** [comment]
Results from rerunning M2243. Note that "Maxlike" in this printout is really actually Maximum A Posteriori (MAP) value. The MAP location is also relatively heavily affected by noise, while the Q50 (median) point should be more robust, given the number of samples I ran yesterday.

Using linear prior on M(<1.5Mpc) & reporting M(<1.5Mpc), as in WtG3: 
mean	1.559574e+15
stddev	2.456749e+14
Q2.5	1.088496e+15
Q25	1.394496e+15
Q50	1.567144e+15
Q75	1.722480e+15
Q97.5	2.031843e+15
HPD68	1.344294e+15	1.830559e+15
HPD95	1.095063e+15	2.035451e+15
MaxLike	1.556060e+15	2.117658e+14	2.744998e+14
Log10 Maxlike	1.520055e+01	7.205249e-02	6.203713e-02
matching reported values, roughly within the numerical accuracy that I ran the MCMC

Using linear prior on M_200 & reporting M_200:
mean	2.456374e+15
stddev	6.296741e+14
Q2.5	1.411988e+15
Q25	2.000570e+15
Q50	2.384420e+15
Q75	2.871408e+15
Q97.5	3.877630e+15
HPD68	1.826829e+15	2.994444e+15
HPD95	1.268129e+15	3.628871e+15
MaxLike	1.960592e+15	1.337623e+14	1.033853e+15
Log10 Maxlike	1.537535e+01	9.083313e-02	1.186523e-01

Various other permutations are possible, but it is important to note that the prior assumed when computing individual cluster masses does make some difference to the reported posterior values. Also, since we don't really care about individual cluster masses, nor are these choices of prior actually the correct prior (which needs a cosmology code to properly include), it's usually best (for simplicity) to run the MCMC s.t. we have a uniform prior in the mass variable we'd like to report. That way it can be thought of as a maximum likelihood scan in that variable.

Note as well from WtG3: ". Finally, we set the
prior on mass, P(M), to be uniform. Note that making the prior
uniform in one measure of mass typically implies a non-uniform
prior in other mass estimates. For this analysis, we measure masses
within an aperture of 1.5 Mpc, and set the prior accordingly."

The prior used was also confirmed with the log files from the MCMC runs used to generate the paper, so that's consistent.

Hope that all makes sense and helps?

**Comment by combet** [comment]
Thanks for running these checks and good to see that the WtGIII mass is recovered!

clusters_mass update: selecting a lin or log mass prior and using or not the WtG STEP2 shear calibration from the config.yaml file has now been enabled, so that there is not need to comment stuff in/out of pzmassfitter to switch from one case to the other when running clusters_mass.

So, using the wtg catalog, linear prior on M_200 & reporting M_200, clusters_mass gives:
mean    5.557732e+14
stddev  3.024005e+14
Q2.5    8.974384e+13
Q25     3.204220e+14
Q50     5.197919e+14
Q75     7.429602e+14
Q97.5   1.204347e+15
HPD68   1.989161e+14    7.771208e+14
HPD95   4.285940e+13    1.145907e+15
MaxLike 4.083457e+14    2.094262e+14    3.690226e+14
Log10 Maxlike   1.502114e+01    5.136452e-01    3.176913e-06

which is ~ factor 4 lower than what is found rerunning the original WtG analysis. More checks underway...

**Comment by nicolaschotard** [comment]
Should this issue be updated according to what has been discovered last week and the DESC meeting?

**Comment by deapplegate** [comment]
The proximate cause for the discrepancy was found: By adding a negative sign to the calculation of the position angle, we were able to bring the two masses into agreement.

However, the root cause is still not completely clear. There's clearly a difference between using RA & DEC, versus using the Xpos and Ypos of galaxies. Need to figure out exactly what is going on.

Part of what led us astray is that the relationship between RA and Xpos in my unit test and Celine's faked image sims was incorrect.

**Comment by combet** [comment]
RA increases to the East and decreases with increasing pixel coordinates (at least that's what looking into DM processed catalogs tells me). Simulations did not have WCS, so I generated fake RADEC from pixel coordinates, but I had both RA and DEC increase with pixel coordinates (x,y). This definition was somehow compatible with the previous position angle sign definition and we could recover the mass for the simulated clusters.

So I checked that having a negative sign for the position angle allows to:
- recover the WTG mass using WTG catalog;
- be compatible with WTG mass using DMstack-produced catalogs;
- recover the input mass using simulated clusters, after I redefined the fake RA to increase in the conventional direction.

I'm closing this issue now, but we should consider revisiting these coordinates definition and maybe switch to using the astropy coordinates library throughout the code instead?
